dmscripts/generate_supplier_user_csv.py

```diff
@@ -0,0 +1,152 @@
+import csv
+from dmutils.s3 import S3ResponseError
+
+from dmscripts.models.writecsv import csv_path
+
+
+def generate_supplier_csv(framework_slug, data_api_client, logger):
+    framework = data_api_client.get_framework(framework_slug).get("frameworks")
+
+    supplier_rows = data_api_client.export_suppliers(framework_slug).get('suppliers', [])
+    if not supplier_rows:
+        logger.info('No supplier data found for framework {}'.format(framework_slug))
+        return [], [], None
+
+    supplier_and_framework_headers = [
+        "supplier_id",
+        "supplier_name",
+        "supplier_organisation_size",
+        "duns_number",
+        "companies_house_number",
+        "registered_name",
+        "declaration_status",
+        "application_status",
+        "application_result",
+        "framework_agreement",
+        "variations_agreed",
+    ]
+    service_count_headers = [lot['slug'] for lot in framework['lots']]
+    contact_info_headers = [
+        'contact_name',
+        'contact_email',
+        'contact_phone_number',
+        'address_first_line',
+        'address_city',
+        'address_postcode',
+        'address_country',
+    ]
+
+    filename = "official-details-for-suppliers-{}".format(framework_slug)
+
+    supplier_headers = supplier_and_framework_headers + ["total_number_of_services"] + \
+        ["service-count-{}".format(h) for h in service_count_headers] + contact_info_headers
+
+    formatted_rows = []
+    for row in supplier_rows:
+        # Include an extra column with the total number of services across all lots
+        service_counts = [row['published_services_count'][heading] for heading in service_count_headers]
+        total_number_of_services = sum(service_counts)
+
+        formatted_rows.append(
+            [row[heading] for heading in supplier_and_framework_headers] +
+            [total_number_of_services] +
+            service_counts +
+            [row['contact_information'][heading] for heading in contact_info_headers]
+        )
+
+    return supplier_headers, formatted_rows, filename
+
+
+def generate_user_csv(framework_slug, data_api_client, user_research_opted_in, logger):
+    """
+    Call API endpoint
+    Get CSV headers and rows
+    """
+    user_headers = [
+        "email address",
+        "user_name",
+        "supplier_id",
+        "declaration_status",
+        "application_status",
+        "application_result",
+        "framework_agreement",
+        "variations_agreed",
+        "published_service_count"
+    ]
+    user_rows = data_api_client.export_users(framework_slug).get('users', [])
+    if not user_rows:
+        logger.info('No user data found for framework {}'.format(framework_slug))
+        return [], [], None
+
+    if user_research_opted_in:
+        user_rows = [row for row in user_rows if row['user_research_opted_in']]
```

Comment:
We can turn this into a generator expression by wrapping in `(` instead of `[` and maybe gain a little performance

Reply:
We can iterate through it just once anyway with a bit of rejigging. I'll push a commit.